src/main.rs

```diff
@@ -0,0 +1,251 @@
+use anyhow::{anyhow, Context, Result};
+use base64::{engine::general_purpose, Engine as _};
+use clap::{Parser, Subcommand};
+use josekit::jwe::alg::direct::DirectJweAlgorithm::Dir;
+use josekit::jwk::Jwk;
+use serde::{Deserialize, Serialize};
+use std::io::{self, Read, Write};
+use std::process::Command as StdCommand;
+use std::thread;
+use std::time::Duration;
+
+#[derive(Debug, Serialize, Deserialize)]
+struct Config {
+    url: Vec<String>,
+    resource_repository: String,
+    resource_type: String,
+}
+
+#[derive(Debug, Serialize, Deserialize)]
+struct ClevisHeader {
+    pin: String,
+    url: Vec<String>,
+    resource_repository: String,
+    resource_type: String,
+    resource_tag: String,
+}
+
+#[derive(Debug, Serialize, Deserialize)]
+pub struct Key {
+    pub key_type: String,
+    pub key: String,
+}
+
+fn fetch_and_prepare_jwk(
+    urls: &[String],
+    resource_repository: &str,
+    resource_type: &str,
+    resource_tag: &str,
+) -> Result<Jwk> {
+    let key = fetch_luks_key(urls, resource_repository, resource_type, resource_tag)?;
+    let key = String::from_utf8(
+        general_purpose::STANDARD
+            .decode(&key)
+            .context("Error decoding key in base64")?,
+    )
+    .context("Error decoding the key in JSON")?;
+    eprintln!("Key: {:?}", key);
+    let key: Key = serde_json::from_str(&key).context("Error in parsing the fetched key")?;
+
+    let mut jwk = Jwk::new(&key.key_type);
+    jwk.set_key_value(&key.key);
+    jwk.set_key_operations(vec!["encrypt", "decrypt"]);
+
+    Ok(jwk)
+}
+
+fn encrypt(config: &str) -> Result<()> {
+    let config: Config =
+        serde_json::from_str(config).map_err(|e| anyhow!("Failed to parse config JSON: {}", e))?;
+
+    let mut input = Vec::new();
+    io::stdin().read_to_end(&mut input)?;
+
+    // TODO: the id needs to be generated for every node
+    let tag = "machine".to_string();
```

Comment:
The tag can also be provided in 'config'

Reply:
this isn't possible because it is the piece that distinguish every secret for each VM and the config needs to remain the same for all the VMs. So, either the entire the path of this element needs to be set it at run time

Reply:
Does the clevis-pin gets/generates the 'tag' or does ignition do it ?